File: 0146-lru-cache/0146-lru-cache.cpp
// struct Node{
//     int key;
//     int val;
//     Node *prev;
//     Node *next;
//     Node(int k, int v, Node* p, Node* n) : key(k),val(v),next(n),prev(p) {}
// };


// class LRUCache {
// private:
//     void deleteNode(Node* node){
//         Node* prevNode=node->prev;
//         Node* afterNode=node->next;
//         prevNode->next=afterNode;
//         afterNode->prev= prevNode;
//     }

//     void insertAhead(Node* node, Node* head){
//         Node* curr=head->next;
//         head->next=node;
//         node->prev=head;
//         node->next=curr;
//         curr->prev=node;

//     }
// public:
//     unordered_map<int,Node*> mpp;
//     int capacity;
//     Node* head=new Node(-1,-1,nullptr,nullptr);
//     Node* tail=new Node(-1,-1,nullptr,nullptr);
    

//     LRUCache(int capacity) {
//         this->capacity=capacity;
//         mpp.clear();
//         head->next=tail;
//         tail->prev=head;
        
//     }
    
//     int get(int key) {

//         if(mpp.find(key) == mpp.end()) return -1;
        
//         Node* node=mpp[key];
//         deleteNode(node);
//         insertAhead(node,head);
//         return node->val;
//     }
    
//     void put(int key, int value) {

//         if(mpp.find(key)!=mpp.end()){
//             Node* node=mpp[key];
//             deleteNode(node);
//             insertAhead(node,head);
//             node->val=value;
//         }else{

//             if(capacity==mpp.size()){
//                 Node* node=tail->prev;
//                 mpp.erase(node->key);
//                 deleteNode(node);
//             }
//             Node* newNode= new Node(key,value,nullptr,nullptr);
//             mpp[key]= newNode;
//             insertAhead(newNode,head);
//         }
        
//     }
// };

// /**
//  * Your LRUCache object will be instantiated and called as such:
//  * LRUCache* obj = new LRUCache(capacity);
//  * int param_1 = obj->get(key);
//  * obj->put(key,value);
//  */

class LRUCache {
private:
    int capacity;
    list<pair<int, int>> lruList; // {key, value}
    unordered_map<int, list<pair<int, int>>::iterator> cache;

public:
    LRUCache(int capacity) {
        this->capacity = capacity;
    }

    int get(int key) {
        // Not found
        if (cache.find(key) == cache.end()) return -1;

        // Found → move to front
        auto it = cache[key];
        int value = it->second;

        lruList.erase(it);
        lruList.push_front({key, value});
        cache[key] = lruList.begin();

        return value;
    }

    void put(int key, int value) {
        // If key exists → remove old
        if (cache.find(key) != cache.end()) {
            lruList.erase(cache[key]);
        }

        // Insert new key-value at front
        lruList.push_front({key, value});
        cache[key] = lruList.begin();

        // If over capacity → remove least recently used (back of list)
        if (cache.size() > capacity) {
            auto last = lruList.back();
            cache.erase(last.first);
            lruList.pop_back();
        }
    }
};

File: 0146-lru-cache/0146-lru-cache.py
class Node:
    def __init__(self, key, val):
        self.key, self.val = key, val
        self.prev = self.next = None
        
class LRUCache:

    def __init__(self, capacity: int):
        self.cap = capacity
        self.cache = {}  # key -> Node

        # Dummy head and tail
        self.head, self.tail = Node(0, 0), Node(0, 0)
        self.head.next, self.tail.prev = self.tail, self.head

    def remove(self, node):
        """Unlink `node` from its current position."""
        prev, nxt = node.prev, node.next
        prev.next, nxt.prev = nxt, prev

    def insert(self, node):
        """
        Insert `node` right after head,
        marking it as the most-recently used.
        """
        nxt = self.head.next
        self.head.next = node
        node.prev = self.head
        node.next = nxt
        nxt.prev = node

    def get(self, key: int) -> int:
        node = self.cache.get(key, None)
        if not node:
            return -1
        # Move to front (most-recent)
        self.remove(node)
        self.insert(node)
        return node.val

    def put(self, key: int, value: int) -> None:
        # If updating, remove old node
        if key in self.cache:
            self.remove(self.cache[key])

        # Insert new node at front
        new_node = Node(key, value)
        self.cache[key] = new_node
        self.insert(new_node)

        # Evict least-recent if over capacity
        if len(self.cache) > self.cap:
            lru = self.tail.prev  # node before tail is the least-recent
            self.remove(lru)
            del self.cache[lru.key]
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)